tests/phpunit/api/v4/Action/SqlExpressionTest.php

```diff
@@ -95,4 +95,26 @@ public function testSelectAlias() {
       ->execute();
   }
 
+  public function testSelectEquations() {
+    Contact::create(FALSE)->addValue('first_name', 'bob')
+      ->execute()->first();
+    $result = Contact::get(FALSE)
+      ->setSelect([
+        'IF((first_name = "bob"), "Yes", "No") AS is_bob',
+        'IF((first_name = "fred"), "Yes", "No") AS is_fred',
+        '(5 * 11)',
+        '(5 > 11) AS five_greater_eleven',
+        '(5 <= 11) AS five_less_eleven',
+      ])
+      ->addWhere('first_name', '=', 'bob')
+      ->setLimit(1)
+      ->execute()
+      ->first();
+    $this->assertEquals('Yes', $result['is_bob']);
+    $this->assertEquals('No', $result['is_fred']);
+    $this->assertEquals('55', $result['5_11']);
```

Comment:
hmm surely it should be `<number>_<op>_<number>`

Reply:
> would the field name for both resolve to '5_11'

Yes, which is why you should always supply your own alias `(5 * 11) AS multiply_5_and_11`

> hmm surely it should be `<number>_<op>_<number>`

APIv4 doesn't allow any non-alphanumeric characters in SQL aliases.

Reply:
@colemanw I'm +1 on merging this now - but I'd like to get a +1 from @seamuslee001 on the security aspects. My understanding is that you parse for a very specific format / range of values? How do you check I'm not adding in a field I don't have access to (eg. api_key)?

Reply:
That's the beauty of the `SqlExpression` classes in APIv4 - I didn't have to start from scratch. I just re-used the same  handling for joining/aliasing/validating/securing fields and other inputs.

So this should be secure from SQLi and privilege escalation, but FYI like the `SqlFunction` classes it doesn't validate that the query will actually work. It won't stop you from e.g. dividing by zero.

Reply:
Some results of `r-run`ning. The examples are in no particular order, but I do group them based on how they executed.

```php
?php
CRM_Core_DAO::executeQuery('UPDATE civicrm_contact SET api_key = NULL WHERE id = 198');
CRM_Core_DAO::executeQuery('UPDATE civicrm_contact SET api_key = "" WHERE id = 199');
CRM_Core_DAO::executeQuery('UPDATE civicrm_contact SET api_key = "asdfasdf" WHERE id = 200');

$v = civicrm_api4('Contact', 'get', [
  'where' => [['id', 'IN', [198,199,200]]],

  // == This works as expected. Visibility for `api_key` and `MAX(api_key)` adjusts properly to permission.
  //'select' => ['id', 'display_name', 'api_key', 'MAX(id)', 'MAX(api_key)'],
  //'groupBy' => ['id'],

  // == This executes. Visibility for `api_key` adjusts properly to permission.
  // == HOWEVER, `has_api_key` does not yield intuitive result. All items say "no". 
  // But maybe that's a MySQL behavior...
  //'select' => ['id', 'display_name', 'api_key', 'IF(api_key, "yes", "no") AS has_api_key'],

  // == All of these variants execute and return the expected value for `has_api_key` (contingent on permission).
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key = ""), "no", "yes") AS has_api_key'], // less useful in null-case
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key != ""), "yes", "no") AS has_api_key'], // more useful in null-case

  // == All of these variants execute. The basic ref to `api_key` adjusts properly to permission.
  // == HOWEVER, the `has_api_key` property is not returned (even for the admin-user with suitable permission).
  // == If the syntax is unsupported, shouldn't it throw an error?
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key IS NULL), "no", "yes") AS has_api_key'],
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key IS ""), "no", "yes") AS has_api_key'],
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key == ""), "no", "yes") AS has_api_key'],
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key <> ""), "yes", "no") AS has_api_key'],
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key IS NOT NULL), "yes", "no") AS has_api_key'],
  //'select' => ['id', 'display_name', 'api_key', 'IF((api_key IS NULL OR api_key = ""), "no", "yes") AS has_api_key'],

  // == All of these throw "Illegal sql expression"
  // 'select' => ['id', 'display_name', 'api_key', 'IF(((api_key IS NULL) OR (api_key = "")), "no", "yes") AS has_api_key'],
  // 'select' => ['id', 'display_name', 'api_key', 'IF( ( (api_key IS NULL) OR (api_key = "") ), "no", "yes") AS has_api_key'],

  // == All of these throw "Incorrect number of arguments for SQL function IF", but
  // I think the consistent with the PR's description (ie parens mandatory).
  //'select' => ['id', 'display_name', 'api_key', 'IF(api_key = "", "no", "yes") AS has_api_key'],
  //'select' => ['id', 'display_name', 'api_key', 'IF(api_key != "", "yes", "no") AS has_api_key'],

]);
print_r((array)$v);
```

I don't see any examples of computations on joined data in the test. (There probably should be some...) But this does appear to work, as in::

```php

$v = civicrm_api4('Email', 'get', [
  'where' => [['contact_id', 'IN', [198,199,200]]],
  'select' => ['contact_id.display_name', 'contact_id.api_key', 'IF((contact_id.api_key != ""), "yes", "no") AS has_api_key'],
]);
```

I fiddled a bit with doing an example that used by groupBy and an `IF()` on the aggregate value. I tried with a few variations on ()'s (because there's the constraint about mandatory ()'s) but couldn't quickly intuit how to the get them working (ie crashed in each case I tried). But it's a niche-thing, and it's OK to have some unsupported+crashy cases.

Reply:
> HOWEVER, the `has_api_key` property is not returned

@totten I think this is because this PR does not add support for the operators you are using (IS, IS NOT, ==, <>), and I think it's swallowing the error because SELECT clauses in APIv4 are forgiving by design... but IMO allowing illegal expressions is a bit too forgiving.

Some of these other cases ought to be unit tests. I'll see about adding them.